### ISSUE TITLE
Cross validate the ABI of sys crates.

### ISSUE THREAD

**Issue by tmiasko** [gtk-rs/gir#555]
It would nice to cross validate our Rust structures with those from C. It could
detect potential defects in Rust generator itself, but also in g-ir-scanner and
consumed GIR XML files.

Implementation-wise one could:

1. Gather C-includes from GIR XML.
2. Generate C program that calculates size and alignments of records from GIR.
3. Compile it with flags from pkg-config.
4. Compare the output with Rust side.

**Comment by sdroege** [comment]
:+1: Good idea! But don't forget to ignore truncated structs :)

**Comment by EPashkin** [comment]
Not sure that it good idea.
But it it really needed https://github.com/rust-lang-nursery/rust-bindgen can be used, and your code can compared with it.

**Comment by sdroege** [comment]
> Not sure that it good idea.

Why not?

**Comment by EPashkin** [comment]
As I know this functionality already present in rust-bingen.

**Comment by sdroege** [comment]
Right, we could build something around rust-bindgen. We should make use of anything that exists out there already.

But it would be a good idea nonetheless to have something for this in place :)

**Comment by tmiasko** [comment]
The rust-bindgen is pretty cool. It indeed does perform layout checks. They have it easy, as they can just ask libclang what is sizeof / alignment of particular type :-).

Though, I want to verify layout of code generated by gir, so rust-bindgen doesn't seem to be immediately applicable here. One could consider code generation that incorporates rust-bindgen in some form, but that's completely different story.  

I wrote a prototype checker, and so far learned that:

* Information about C headers in GIR XML is incomplete. For example Gtk doesn't include any headers.
* Some types are visible conditionally based on macro definitions.
* Gio includes parts that are specific to UNIX, with separate pkg-config, but you can't easily tell them apart by looking at XML.

Prototype generates a single C file that calculates sizeof and alignment of all complete types from GIR. That turned out to be a bad idea. A single issue makes everything fail to compile.

Doing separate compilation step for each type would be more resilient to above problem, and could provide valuable feedback despite partial failures. That is what I will try next.

**Comment by tmiasko** [comment]
In gdk-sys there is ABI mismatch for GdkEventTouchpadPinch and GdkEventTouchpadSwipe.
Both of them have phase field where actual type is gint8, but is overridden
with an annotation to GdkTouchpadGesturePhase. The underlying type of the enumeration 
is an int, and so there is mismatch.

For one, this is certainly some form of bug in generated gir data, because it
doesn't even retain c:type that would allow you to calculate actual size.  One
option then would be to retain it. At this point I am not sure we would
interpret this correctly, nor if other consumers of gir would. Thus one could
consider another solution: get rid of type annotation completely.

**Comment by EPashkin** [comment]
@tmiasko Thanks for find.
If you meant just adding ctype manually, I agree.
We can patch .gir file
```diff
--- a/Gdk-3.0.gir
+++ b/Gdk-3.0.gir
@@ -6993,7 +6993,7 @@ gdk_event_get_source_device() to get the slave device.</doc>
       </field>
       <field name="phase" writable="1">
         <doc xml:space="preserve">the current phase of the gesture</doc>
-        <type name="TouchpadGesturePhase"/>
+        <type name="TouchpadGesturePhase" c:type="gint8"/>
       </field>
       <field name="n_fingers" writable="1">
         <doc xml:space="preserve">The number of fingers triggering the pinch</doc>

```
and it produces
```diff
--- a/gdk-sys/src/lib.rs
+++ b/gdk-sys/src/lib.rs
 #![allow(non_camel_case_types, non_upper_case_globals, non_snake_case)]
@@ -3675,7 +3675,7 @@ pub struct GdkEventTouchpadPinch {
     pub type_: GdkEventType,
     pub window: *mut GdkWindow,
     pub send_event: i8,
-    pub phase: GdkTouchpadGesturePhase,
+    pub phase: i8,
     pub n_fingers: i8,
     pub time: u32,
     pub x: c_double,
```
This not problem for us as we usually make getter in manual way, just need don't forget about actual enum when do getter.

**Comment by tmiasko** [comment]
Adding c:type would work for Rust. Unfortunately this is not the case for
languages which generate bindings based on typelib, because it doesn't contain
information about c:type. Thus, while reporting issue upstream [0], I suggested
different solution: remove the type annotation.

If maintaining manual fix to gir xml is not too burdensome, we could do that as well,
of course.

[0] https://gitlab.gnome.org/GNOME/gobject-introspection/issues/193

**Comment by EPashkin** [comment]
By removing type annotation you meant make it `<type c:type="gint8"/>`?
It need more edit action ;)
Added fix to gir-files https://github.com/gtk-rs/gir-files/pull/15

**Comment by tmiasko** [comment]
Without annotation it would look like: `<type name="gint8" c:type="gint8"/>`.

**Comment by tmiasko** [comment]
Now that basic layout tests are in PR #558, I was looking at comparing constants as well. 

My initial ideas were as follows, but maybe you have others (I wanted to avoid type specific formatting in case type is wrong in XML):
* Applying stringifying macro in C and comparing it with Rust - this is super fast, but you might get things like `(1 + 2)` back, instead of `3`.
* Using C++ and streams to print things out - very convenient, but quite slow to compile with more than 2000 constants in Gdk.

There are some things that are generally expected to differ like:
* ..._BINARY_AGE
* ..._MINOR_VERSION
* ..._MICRO_VERSION, 
* ..._VERSION_MIN_REQUIRED

But also some actual differences:
```
Constant value mismatch for G_LOG_FATAL_MASK
Rust: "0"
C:    "5"
Constant value mismatch for G_ANALYZER_ANALYZING
Rust: "1"
C:    "0"
Constant value mismatch for G_SEARCHPATH_SEPARATOR_S
Rust: ";"
C:    ":"
Constant value mismatch for GDK_PRIORITY_REDRAW
Rust: "20"
C:    "120
```

**Comment by EPashkin** [comment]
Yes, this problem, but it IMHO need be checked in gobject introspection, not in gir, while you can help to find constant difference gobject introspection and report bugs.
From that i see gobject introspection bad with expression macros:
```c
  G_LOG_FLAG_RECURSION          = 1 << 0,
  G_LOG_LEVEL_ERROR             = 1 << 2,       /* always fatal */
#define G_LOG_FATAL_MASK        (G_LOG_FLAG_RECURSION | G_LOG_LEVEL_ERROR)

#define G_PRIORITY_HIGH_IDLE        100
#define GDK_PRIORITY_REDRAW     (G_PRIORITY_HIGH_IDLE + 20)
```
G_SEARCHPATH_SEPARATOR_S is platform-depend and we not publish it from glib.
G_ANALYZER_ANALYZING seems internal constant

**Comment by tmiasko** [comment]
That is true that all issues detected will be essentially in gobject-introspection 
not in Rust gir generator. But I still think it is worthwhile to do, because
no one is checking that in gobject-introspection, and any bugs will affect Rust 
bindings either way.

Generating tests in sys crates makes such test trivial to execute. 
Any portability problems on different operating systems and targets
will be detected right away. And you can do all that without manual review.

* G_SEARCHPATH_SEPARATOR_S, is not fixed just yet :-) https://github.com/gtk-rs/sys/pull/81
* G_LOG_FATAL_MASK should be fixed in next gobject-introspection release (see https://gitlab.gnome.org/GNOME/gobject-introspection/issues/173), situation with GDK_PRIORITY_REDRAW might have improved as well.
* G_ANALYZER_ANALYZING - indeed, macro definitions without any docs probably shouldn't be in gir.

**Comment by EPashkin** [comment]
For G_SEARCHPATH_SEPARATOR_S i meant https://github.com/gtk-rs/glib/pull/296/files#diff-e3845858ec246b527c915f90e6f8ae87R227

**Comment by sdroege** [comment]
The version constants also seem kind of useless and should be ignored

**Comment by tmiasko** [comment]
Implemented in #558 and #565.
